caas/kubernetes/clientconfig/plugins_test.go

```diff
@@ -0,0 +1,326 @@
+			},
+			{
+				NonResourceURLs: []string{rbacv1.NonResourceAll},
+				Verbs:           []string{rbacv1.VerbAll},
+			},
+		},
+	}
+
+	clusterRoleBinding := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: clientconfig.JujuClusterRoleBindingName,
+		},
+		RoleRef: rbacv1.RoleRef{
+			Kind: "ClusterRole",
+			Name: cr.Name,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      rbacv1.ServiceAccountKind,
+				Name:      saName,
+				Namespace: s.namespace,
+			},
+		},
+	}
+
+	gomock.InOrder(
+		s.mockClusterRoles.EXPECT().Get(cr.Name, metav1.GetOptions{}).Times(1).
+			Return(cr, nil),
+		s.mockServiceAccounts.EXPECT().Create(newSa).Times(1).
+			Return(newSa, nil),
+		s.mockServiceAccounts.EXPECT().Get(newSa.Name, metav1.GetOptions{}).Times(1).
+			Return(newSa, nil),
+		s.mockClusterRoleBinding.EXPECT().Create(clusterRoleBinding).Times(1).
+			Return(clusterRoleBinding, nil),
+		s.mockServiceAccounts.EXPECT().Create(newSa).Times(1).
+			Return(nil, s.k8sAlreadyExistsError()),
+		s.mockServiceAccounts.EXPECT().Get(newSa.Name, metav1.GetOptions{}).Times(1).
+			Return(saWithSecret, nil),
+		s.mockSecrets.EXPECT().Get(saWithSecret.Secrets[0].Name, metav1.GetOptions{}).Times(1).
+			Return(secret, nil),
+	)
+
+	cfgOut, err := clientconfig.EnsureJujuAdminServiceAccount(s.k8sClient, cfg, contextName)
+	c.Assert(err, jc.ErrorIsNil)
+	authName := cfg.Contexts[contextName].AuthInfo
+	updatedAuthInfo := cfgOut.AuthInfos[authName]
+	c.Assert(updatedAuthInfo.AuthProvider, gc.IsNil)
+	c.Assert(updatedAuthInfo.ClientCertificateData, gc.DeepEquals, secret.Data[core.ServiceAccountRootCAKey])
+	c.Assert(updatedAuthInfo.Token, gc.Equals, string(secret.Data[core.ServiceAccountTokenKey]))
+
+}
+
+func (s *K8sRawClientSuite) TestEnsureClusterRole(c *gc.C) {
+	s.ctrl = s.setupBroker(c)
+	defer s.ctrl.Finish()
+
+	cr := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "juju-admin-cluster-role",
+			Namespace: s.namespace,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{rbacv1.APIGroupAll},
+				Resources: []string{rbacv1.ResourceAll},
+				Verbs:     []string{rbacv1.VerbAll},
+			},
+			{
+				NonResourceURLs: []string{rbacv1.NonResourceAll},
+				Verbs:           []string{rbacv1.VerbAll},
+			},
+		},
+	}
+
+	gomock.InOrder(
+		s.mockClusterRoles.EXPECT().Get(cr.Name, metav1.GetOptions{}).Times(1).
+			Return(cr, nil),
+	)
+	crOut, err := clientconfig.EnsureClusterRole(s.k8sClient, cr.Name, s.namespace)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(crOut, jc.DeepEquals, cr)
+
+	gomock.InOrder(
+		s.mockClusterRoles.EXPECT().Get(cr.Name, metav1.GetOptions{}).Times(1).
+			Return(nil, s.k8sNotFoundError()),
+		s.mockClusterRoles.EXPECT().Create(cr).Times(1).
+			Return(cr, nil),
+	)
+	crOut, err = clientconfig.EnsureClusterRole(s.k8sClient, cr.Name, s.namespace)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(crOut, jc.DeepEquals, cr)
+
+}
+
+func (s *K8sRawClientSuite) TestEnsureServiceAccount(c *gc.C) {
```

Comment:
Generally we don't bother writing unit tests for private methods, unless they have important and/or complex business logic. I think all we need is to set up a few scenarios for ensureJujuServiceAdminAccount(). We could simply just have 2 tests:
TestEnsureJujuServiceAdminAccount()
TestEnsureJujuServiceAdminAccountIdempotent()

The second test would call ensureJujuServiceAdminAccount() twice and check that the end result is the same and the expected mock calls are made.

Reply:
yes, add one more TestEnsureJujuServiceAdminAccountIdempotent test.
Do I need to remove the other tests or just leave it(more tests don't hurt)?

Reply:
added TestEnsureJujuServiceAdminAccountIdempotent for testing after 1st run.